auth/src/source.rs

```diff
@@ -0,0 +1,474 @@
+            .send()
+            .await
+            .map_err(|_| Error::Other("unable to make request to oauth endpoint".into()))?;
+        if !res.status().is_success() {
+            return Err(Error::Other(format!(
+                "bad request with status: {}",
+                res.status()
+            )));
+        }
+        let token_response: TokenResponse = res
+            .json()
+            .await
+            .map_err(|_| Error::Other("unable to decode response".into()))?;
+
+        Ok(AccessToken {
+            value: token_response.access_token,
+            expires: Some(Utc::now() + Duration::seconds(token_response.expires_in)),
+        })
+    }
+
+    fn signer(&self) -> Result<Box<dyn Signer>> {
+        let pk = rustls_pemfile::read_one(&mut self.file.private_key.as_bytes())?
+            .ok_or_else(|| Error::Other("".into()))?;
+        let pk = match pk {
+            Item::RSAKey(item) => item,
+            Item::PKCS8Key(item) => item,
+            _ => return Err(Error::Other("expected key in a different format".into())),
+        };
+        rustls::sign::RsaSigningKey::new(&rustls::PrivateKey(pk))
+            .map_err(|_| Error::Other("unable to create signer".into()))?
+            .choose_scheme(&[rustls::SignatureScheme::RSA_PKCS1_SHA256])
+            .ok_or_else(|| Error::Other("invalid signing scheme".into()))
+    }
+
+    fn payload(&self, signer: Box<dyn Signer>) -> Result<String> {
+        let scopes = self.scopes.join(" ");
+        let mut claims = JwsClaims {
+            iss: self.file.client_email.as_str(),
+            scope: Some(scopes.as_str()),
+            aud: self.file.token_uri.as_str(),
+            exp: None,
+            iat: None,
+            sub: None,
+            typ: None,
+        };
+        let header = DEFAULT_HEADER;
+
+        let ss = format!("{}.{}", header.encode()?, claims.encode()?);
+        let sig = signer
+            .sign(ss.as_bytes())
+            .map_err(|_| Error::Other("unable to sign bytes".into()))?;
+        Ok(format!(
+            "{}.{}",
+            ss,
+            base64::encode_config(sig, base64::URL_SAFE_NO_PAD)
+        ))
+    }
+}
+
+#[async_trait]
+impl Source for ServiceAccountKeySource {
+    async fn token(&self) -> Result<AccessToken> {
+        self._access_token().await
+    }
+}
+
+/// The request body of a Service Account Key token exchange.
+#[derive(Serialize)]
+struct ServiceAccountTokenRequest {
+    grant_type: String,
+    assertion: String,
+}
+
+/// The response of a Service Account Key token exchange.
+#[derive(Deserialize)]
+struct TokenResponse {
+    access_token: String,
+    token_type: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    id_token: Option<String>,
+    expires_in: i64,
+}
+
+/// Options for building a [UserSource].
+pub struct UserSourceBuilder<'a> {
+    file_path: Option<PathBuf>,
+    contents: Option<&'a [u8]>,
+
+    scopes: Vec<String>,
+}
+
+impl UserSourceBuilder<'_> {
```

Comment:
Apply my comments from ServiceAccountKeySource to the other sources.

Reply:
done.